File: Java/Ema/Src/main/java/com/thomsonreuters/ema/access/VectorEntry.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2015. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;

import java.nio.ByteBuffer;

/**
 * VectorEntry represents an entry of Vector.
 * <br>VectorEntry associates entry's position, action, permission information, data and its data type.
 */
public interface VectorEntry
{
	/**
	 * VectorAction represents vector entry action.
	 */
	public static class VectorAction
	{
		/**
		 * Indicates a partial change of the entry.
		 */
		public final static int UPDATE	= 1;
	
		/**
		 * Indicates to replace the entry.
		 */
		public final static int SET		= 2;
	
		/**
		 * Indicates to empty the entry. Contains no data.
		 */
		public final static int CLEAR	= 3;
	
		/**
		 * Indicates to place the entry in between other entries.
		 * <br>Increases any higher-ordered position by one.
		 * <br>May leave gaps if previous lower-ordered position is unpopulated. 
		 */
		public final static int INSERT	= 4;
	
		/**
		 * Indicates to remove the entry.
		 * <br>Decreases any higher-ordered position by one.
		 * <br>Contains no data.
		 */
		public final static int DELETE	= 5;
	}
	
	/**
	 * Returns the VectorAction value as a string format.
	 * 
	 * @return String containing string representation of VectorAction
	 */
	public String vectorActionAsString();
		
	/**
	 * Returns the DataType of the entry's load.
	 * <br>Return of {@link com.thomsonreuters.ema.access.DataType.DataTypes#NO_DATA} signifies no data present in load.
	 * <br>Return of {@link com.thomsonreuters.ema.access.DataType.DataTypes#ERROR} signifies error while extracting content of load.
	 * 
	 * @return data type of the contained object
	 */
	public int loadType();
	
	/**
	 * Returns a string representation of the class instance.
	 * 
	 * @return string representation of the class instance
	 */
	@Override
	public String toString();
	
	/**
	 * Indicates presence of PermissionData.
	 * 
	 * @return true if permission data is set; false otherwise
	 */
	public boolean hasPermissionData();
	
	/**
	 * Returns position of the entry.
	 * 
	 * @return entry's position
	 */
	public long position();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.ReqMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.ReqMsg} class reference to contained object
	 */
	public ReqMsg reqMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.RefreshMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.RefreshMsg} class reference to contained entry's load object
	 */
	public RefreshMsg refreshMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.UpdateMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.UpdateMsg} class reference to contained entry's load object
	 */
	public UpdateMsg updateMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.StatusMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.StatusMsg} class reference to contained entry's load object
	 */
	public StatusMsg statusMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.PostMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.PostMsg} class reference to contained entry's load object
	 */
	public PostMsg postMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.AckMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.AckMsg} class reference to contained entry's load object
	 */
	public AckMsg ackMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.GenericMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.GenericMsg} class reference to contained entry's load object
	 */
	public GenericMsg genericMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.FieldList}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.FieldList} class reference to contained entry's load object
	 */
	public FieldList fieldList();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.ElementList}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.ElementList} class reference to contained entry's load object
	 */
	public ElementList elementList();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.Map}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.Map} class reference to contained entry's load object
	 */
	public Map map();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.Vector}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.Vector} class reference to contained entry's load object
	 */
	public Vector vector();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.Series}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.Series} class reference to contained entry's load object
	 */
	public Series series();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.FilterList}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.FilterList} class reference to contained entry's load object
	 */
	public FilterList filterList();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmOpaque}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmOpaque} class reference to contained entry's load object
	 */
	public OmmOpaque opaque();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmXml}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmXml} class reference to contained entry's load object
	 */
	public OmmXml xml();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmAnsiPage}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmAnsiPage} class reference to contained entry's load object
	 */
	public OmmAnsiPage ansiPage();
	
	/**
	 * Returns the contained Data based on the DataType.
	 * 
	 * @return Data class reference to contained entry's load object
	 */
	public Data load();
	
	/**
	 * Returns the current action on the OMM data.
	 * 
	 * @return VectorEntry Action
	 */
	public int action();
	
	/**
	 * Returns PermissionData.
	 * 
	 * @throw OmmInvalidUsageException if hasPermissionData() returns false
	 * 
	 * @return ByteBuffer containing permission information
	 */
	public ByteBuffer permissionData();
	
	/**
	 * Returns Error.
	 * 
	 * @throw OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmError}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmError} class reference to contained entry's load object
	 */
	public OmmError error();
}

File: CSharp/Ema/Src/Core/Access/VectorEntry.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Codec;
using LSEG.Eta.Common;
using System;

namespace LSEG.Ema.Access
{
    /// <summary>
    /// VectorEntry represents an entry of Vector.<br/>
    /// VectorEntry associates entry's position, action, permission information, data and its data type.
    /// </summary>
    /// <remarks>
    /// Objects of this class are intended to be short lived or rather transitional.<br/>
    /// This class is designed to efficiently perform setting and extracting of Vector and its content.<br/>
    /// Objects of this class are not cache-able.<br/>
    /// </remarks>
    public sealed class VectorEntry : Entry
    {
        internal Eta.Codec.VectorEntry m_rsslVectorEntry;
        private EmaBuffer m_permissionBuffer = new EmaBuffer();

        /// <summary>
        /// Indicates presence of PermissionData.
        /// </summary>
        public bool HasPermissionData { get => m_rsslVectorEntry.CheckHasPermData(); }

        internal VectorEntry() 
        {
            m_rsslVectorEntry = new Eta.Codec.VectorEntry();
        }
        
        /// <summary>
        /// Returns the VectorAction value as a string format.
        /// </summary>
        /// <returns>string representing VectorEntry action</returns>
        public string VectorActionAsString() { return VectorAction.VectorActionToString(Action); }
        
        /// <summary>
        /// Returns position of the entry
        /// </summary>
        /// <returns>entry's position</returns>
        public long Position { get => m_rsslVectorEntry.Index; }
        
        /// <summary>
        /// Returns the current action on the OMM data.
        /// </summary>
        /// <returns>VectorEntry Action</returns>
        public int Action { get => VectorAction.EtaVectorActionToInt(m_rsslVectorEntry.Action); }

        /// <summary>
        /// Returns PermissionData.
        /// </summary>
        /// <returns><see cref="EmaBuffer"/> containing permission information</returns>
        /// <exception cref="OmmInvalidUsageException">Thrown if <see cref="HasPermissionData"/> returns false</exception>
        public EmaBuffer PermissionData 
        {
            get
            {
                if (!m_rsslVectorEntry.CheckHasPermData())
                {
                    throw new OmmInvalidUsageException("Attempt to access Permission Data while VectorEntry doesn't have it.");
                }
                m_permissionBuffer.Clear();
                var span = new Span<byte>(m_rsslVectorEntry.PermData.Data().Contents, m_rsslVectorEntry.PermData.Position, m_rsslVectorEntry.PermData.Length);
                m_permissionBuffer.Append(span);
                return m_permissionBuffer;
            }
        }

        internal CodecReturnCode Decode(DecodeIterator decodeIterator)
        {
            m_rsslVectorEntry.Clear();
            return m_rsslVectorEntry.Decode(decodeIterator);
        }

        /// <summary>
        /// Clears current VectorEntry instance
        /// </summary>
        public void Clear()
        {
            m_rsslVectorEntry.Clear();
            m_permissionBuffer.Clear();
            ClearLoad();
        }

        /// <summary>
        /// Provides string representation of the current instance
        /// </summary>
        /// <returns>string representing current <see cref="VectorEntry"/> object.</returns>
        public override string ToString()
        {
            if (Load == null)
                return $"{NewLine}ToString() method could not be used for just encoded object.{NewLine}";

            m_toString.Length = 0;
            m_toString.Append("VectorEntry ")
                    .Append(" action=\"").Append(VectorActionAsString()).Append("\"")
                    .Append(" index=\"").Append(Position);

            if (HasPermissionData)
            {
                m_toString.Append("\" permissionData=\"").Append(PermissionData).Append("\"");
                Utilities.AsHexString(m_toString, PermissionData).Append("\"");
            }

            m_toString.Append("\" dataType=\"").Append(DataType.AsString(Load.DataType)).Append("\"").AppendLine();
            m_toString.Append(Load.ToString(1));
            Utilities.AddIndent(m_toString, 0).Append("VectorEntryEnd").AppendLine();

            return m_toString.ToString();
        }
    }
}
